userguide/payment/includes/payment-methods-flows.adoc

```diff
@@ -0,0 +1,93 @@
+
+* *Server-side tokenization* - In this case, the end user's payment details are sent to Kill Bill which invokes the payment gateway to perform tokenization.
+
+The flow diagrams below explain these payment method flows. We consider the following actors:
+
+* *Browser*: The end-user sitting behind a browser and initiating the payment method flow
+
+* *Merchant Site*: customer facing web site which receives the order
+
+* *Kill Bill* - The Kill Bill system
+
+* *Checkout Servlet* - Servlet that initiates setting up the payment method. 
+
+* *Payment Plugin*: https://docs.killbill.io/latest/payment_plugin.html[Kill Bill plugin] corresponding to the payment gateway 
+
+* *Payment Gateway*: Entity that processes the payment
+
+=== Client-side Tokenization Payment Method Flow
+
+In this case, the merchant website collects the end-user's payment details. These are then converted into a *token* via a payment gateway specific Javascript which is embedded into the merchant site payment page. This token is sent to Kill Bill. Kill Bill then saves it in the Kill Bill database and uses it to charge the customer.
+
+The following diagram explains this payment method flow:
+
+image:https://github.com/killbill/killbill-docs/raw/v3/userguide/assets/img/payment-userguide/clientside-tokenization-payment-method-flow.png[align=center]
+
+. The end user enters payment details on the merchant website.
+
+. The merchant website converts it into a token and invokes the `addPaymentMethod` in Kill Bill with this token. This saves the token in the Kill Bill database.
+
+. Some payment gateways require the token to be saved within the gateway so as to be able to process recurring payments. In such cases, Kill Bill invokes the `addPaymentMethod` in the payment plugin corresponding to the payment gateway. 
+
+. The payment plugin then makes a gateway specific call to save the payment token within the gateway.
+
+=== Hosted Payment Page Payment Method Flow
+
+Sometimes, a merchant site may not provide a payment form/may not want to collect payment data and tokenize it. The hosted payment page flow is useful in such situations. In such cases, the merchant site needs to perform a two-step process. In the first step, it needs to invoke Kill Bill to obtain the URL to redirect the user and redirect the user to this page. Once the end-user enters his/her payment details on this page, the merchant site needs to invoke Kill Bill to add the payment method. Kill Bill then retrieves the payment details from the payment gateway and saves them in the Kill Bill database. This information is then used by Kill Bill to charge the customer.
+
+
+The following diagram explains this payment method flow:
+
+image:https://github.com/killbill/killbill-docs/raw/v3/userguide/assets/img/payment-userguide/hpp-payment-method-flow.png[align=center]
+
+. The end user initiates the checkout process.
+
+. The merchant website invokes a checkout servlet in Kill Bill. 
+
```

Comment:
The checkout servlet lives in the Kill Bill payment plugin (and not really in Kill Bill core). Technically Kill Bill is both the core + plugins so this is correct, but I am not sure if this will be clear for readers.

Reply:
@sbrossie , I've now used a different color for the Checkout Servlet and Payment plugin.  Please let me know if this helps, if not I can draw some boxes to indicate boundaries.